### PR TITLE
[Testing][vLLM] Add vLLM Eagle3 Test Cases

### PR DESCRIPTION
**Pull request by dsikka**
- Test entrypoints with speculators models

### CODE REVIEW

File: tests/e2e/test_vllm.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    @pytest.mark.parametrize(
+        "model_path",
+        [
+            "nm-testing/SpeculatorLlama3-1-8B-Eagle3-converted-0717-quantized",
+            "nm-testing/Speculator-Qwen3-8B-Eagle3-converted-071-quantized",
+        ],
+    )
+    def test_vllm_engine_eagle3(self, model_path):
+        output = self._run_vllm_engine(model_path=model_path)
+        print(output)  # noqa: T201
```
Comment (rahul-tuli) [inline tests/e2e/test_vllm.py]: Let's remove print and log the outputs instead?

File: tests/e2e/test_vllm.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    vllm_installed = True
+except ImportError:
+    vllm_installed = False
+    logger.warning("vllm is not installed. This test will be skipped")
+
+
+@pytest.mark.skipif(not vllm_installed, reason="vLLM is not installed, skipping test")
+class TestEagle3vLLM:
+    # move boilerplate code to some base testing class
```
Comment (rahul-tuli) [inline tests/e2e/test_vllm.py]: nit: we can remove this comment

**Review by rahul-tuli — COMMENTED** [review]
Diff looks good, what are your thoughts on adding vllm installable, like `pip install "speculators[dev, vllm]"`?

**Review by rahul-tuli — APPROVED** [review]

File: tests/e2e/test_vllm.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import pytest
+import torch
+from loguru import logger
+
+try:
```
Comment (markurtz) [inline tests/e2e/test_vllm.py:5]: Replace this try catch with the built in pytest method for this:
`pytest.importorskip("vllm", reason="vLLM is not installed")`


If vllm isn't imported, it will automatically skip all tests within the module. After that, you can use `from vllm import LLM, SamplingParams` normally without having to indent and without having to mark the later tests

File: tests/e2e/test_vllm.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
+import pytest
```
Comment (markurtz) [inline tests/e2e/test_vllm.py:1]: Let's restructure this to either `tests/e2e/vllm/test_eagle3.py` or `tests/e2e/eagle3/test_vllm.py` to make it clearer exactly what this is testing and not over complicate the scope of the each individual test module in the future as more are added.

File: tests/e2e/test_vllm.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    @pytest.mark.smoke
+    @pytest.mark.parametrize(
+        "model_path",
+        [
+            "nm-testing/SpeculatorLlama3-1-8B-Eagle3-converted-0717-quantized",
+            "nm-testing/Speculator-Qwen3-8B-Eagle3-converted-071-quantized",
+        ],
+    )
+    def test_vllm_engine_eagle3(self, model_path):
+        output = self._run_vllm_engine(model_path=model_path)
```
Comment (markurtz) [inline tests/e2e/test_vllm.py:52]: Calling into the helper function is going to hide any specific exceptions, asserts, etc within due to issues with pytest being able to get context from locals within another function. Given the code is minimal, at a minimum the generate should be moved to the test function. For construction, see the comment on the fixture

File: tests/e2e/test_vllm.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        logger.info(output)
+        assert output
```
Comment (markurtz) [inline tests/e2e/test_vllm.py:54]: At a minimum we should at least assert the shape is correct and simple checks on the output values that they are not 0's, NANs, etc

File: tests/e2e/test_vllm.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    def temp_cache_dir(self, tmp_path, monkeypatch):
+        cache_dir = tmp_path / "hf_cache"
+        cache_dir.mkdir(exist_ok=True)
+        monkeypatch.setenv("HF_HOME", str(cache_dir))
+        monkeypatch.setenv("TRANSFORMERS_CACHE", str(cache_dir))
+        monkeypatch.setenv("HUGGINGFACE_HUB_CACHE", str(cache_dir))
+        return cache_dir
+
+    def _run_vllm_engine(self, model_path):
+        sampling_params = SamplingParams(temperature=0.80, top_p=0.95)
```
Comment (markurtz) [inline tests/e2e/test_vllm.py:39]: This should have a seed set to better fix reproducibility and stability of the test

File: tests/e2e/test_vllm.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        llm = LLM(model=model_path, gpu_memory_utilization=0.9)
+        return llm.generate(self.prompts, sampling_params)
```
Comment (markurtz) [inline tests/e2e/test_vllm.py:41]: Generation should be capped with max_tokens to limit the length of the tests, add predictability to them, and make them more readable/debuggable

File: tests/e2e/test_vllm.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+
+@pytest.mark.skipif(not vllm_installed, reason="vLLM is not installed, skipping test")
+class TestEagle3vLLM:
+    def setup_method(self):
+        import gc
+
+        gc.collect()
+        if torch.cuda.is_available():
+            torch.cuda.empty_cache()
```
Comment (markurtz) [inline tests/e2e/test_vllm.py:21]: synchronize should be called before to better ensure cache is reliably emptied

File: tests/e2e/test_vllm.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    @pytest.mark.parametrize(
+        "model_path",
+        [
+            "nm-testing/SpeculatorLlama3-1-8B-Eagle3-converted-0717-quantized",
+            "nm-testing/Speculator-Qwen3-8B-Eagle3-converted-071-quantized",
+        ],
+    )
+    def test_vllm_engine_eagle3(self, model_path):
+        output = self._run_vllm_engine(model_path=model_path)
+        logger.info(output)
```
Comment (markurtz) [inline tests/e2e/test_vllm.py:53]: This can be a very long and noisy output, we should log out a more controlled set here

File: tests/e2e/test_vllm.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+class TestEagle3vLLM:
+    def setup_method(self):
+        import gc
+
+        gc.collect()
+        if torch.cuda.is_available():
+            torch.cuda.empty_cache()
+
+        self.prompts = [
+            "The capital of France is",
```
Comment (markurtz) [inline tests/e2e/test_vllm.py:24]: Utilize more deterministic prompts that don't rely on specific knowledge that may change across models. For example, "Output the following sentence: vLLM is the best inference engine for speculative decoding"

File: tests/e2e/test_vllm.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+            "The president of the US is",
+            "My name is",
+        ]
+
+    @pytest.fixture
+    def temp_cache_dir(self, tmp_path, monkeypatch):
```
Comment (markurtz) [inline tests/e2e/test_vllm.py:30]: This fixture is a no op in the current code since it isn't used anywhere. It needs to either be included in the specific test case later or it needs to be decorated with @pytest.fixture(autouse=True).

But, to future-proof these tests / make them more maintainable as well as removing the later mentioned issue around helper functions, this would be much better if we replace it with a parameterized yield style fixture which handles setup, vLLM instantiation, and teardown like the following:

@pytest.fixture(params=[...model_ids])
def llm_engine(request, tmp_path, monkeypatch):
    model_id = request.param

    # Ensure clean and temporary state
    _clean_cuda()
    cache_dir = tmp_path / "hf_cache"
    cache_dir.mkdir(exist_ok=True)
    for env_key in ("HF_HOME", "TRANSFORMERS_CACHE", "HUGGINGFACE_HUB_CACHE"):
        monkeypatch.setenv(env_key, str(cache_dir))
    monkeypatch.setenv("HF_HUB_DISABLE_TELEMETRY", "1")
    monkeypatch.setenv("HF_HUB_ENABLE_HF_TRANSFER", "1")

    # Construct LLM and yield
    try:
        llm = LLM(model=model_id, gpu_memory_utilization=0.9)

        yield llm, model_id
    finally:
        # Ensure clean ending state
        if llm is not None:
            try:
                del llm
            except Exception as exc:
                logger.warning(f"del llm raised: {exc}")

        _clean_cuda()
